JWT 기반 일반 로그인 / 회원가입 구현

dlehdhun opened the pull request:
## #️⃣ 연관된 이슈
- #5 

## 📝 작업 내용
> 로그인 구현
> 회원가입 구현
> AccessToken 재발급
- AccessToken: body 로 반환
- RefreshToken: 쿠키로 전달되며 Redis 저장 후 검증에 활용
## 💬 리뷰 요구사항
> ValidationPipe 및 ExceptionFilter는 현재 @yxxjxxeee님이 작성 중이라, 에러 반환 부분은 NestJS 기본 데코레이터 반환으로 
임시 처리했습니다.

dlehdhun committed:
feat: (#5) 회원가입 구현

dlehdhun committed:
feat: (#5) 로그인 구현

dlehdhun committed:
feat: (#5) accessToken 재발급 및 로그아웃 구현

- /auth/refresh API 생성
- refreshToken 기반 accessToken 재발급 로직 추가
- Redis 기반 refresh 토큰 검증
- 로그아웃 시 Redis 토큰 삭제 및 쿠키 삭제 기능 추가

dlehdhun committed:
refactor: (#5) 사용자 모델·인증 흐름 정리 및 토큰 처리 개선

- 사용자 모델의 nickname DB 길이를 30으로 수정
- any 타입을 명시적으로 사용하여 타입 안정성 강화
- 인증 흐름의 오류 처리 로직 개선 및 코드 최적화
- ESLint 오류 수정

rrgks6221 commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,50 @@
+export class AuthController {
+  constructor(
+    private readonly authService: AuthService,
+    private readonly cookieService: CookieService,
```
response 객체를 필요로한다는 부분을 봤을 때 서비스라고 칭하기는 어려울거같은데 코드 응집을 위해서라면 authController에 private method로만 관리하는건 어떻게 생각하실까요??
아니면 동일하게 cookie 관련된 로직이 재활용될 여지가 있을까요??

dlehdhun replied:
쿠키 로직을 캡슐화하고 재사용성을 높이기 위해 서비스로 분리해 두었습니다. 현재 계획상 cookie 관련 로직이 추가될 가능성이 높지 않아서 private method로만 관리하는것도 좋은 방법 같습니다!

rrgks6221 commented on src/users/users.repository.ts:
```diff
@@ -0,0 +1,33 @@
+  constructor(private readonly prisma: PrismaService) {}
+
+  async exists(where: Prisma.UserWhereUniqueInput): Promise<boolean> {
+    const count = await this.prisma.user.count({ where });
```
잘 기억은 안나는데 prisma 제공 메서드중에 exsit라는 메서드가 있었던거같은데 있을까요?

dlehdhun replied:
Prisma에 exist라는 메서드가 있었던 것 같은데, 구버전에서 쓰이던 문법이라 지금은 지원되지 않는 것 같아요!

rrgks6221 replied:
아 구시절 유물이 됐군요
실제로 exists랑 count는 DB에서 동작 차이가 있기때문에 exists가 더 성능면에서 좋은데 아쉽군요

count 메서드로 exists를 제공해야한다면 where 조건을 줄이는건 어떨까요??
사용처에서 where 조건이 느슨하게 된다면 모든 row를 탐색해서 집계하기 떄문에 부하가 있을 수 있어요

존재여부를 확인할때는 보통 index가 존재하는 컬럼으로 많이 합니다. (PK, FK, 또다른 인덱스 컬럼)

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  // 비밀번호 확인용 유저 조회
+  async findUserForAuth(username: string): Promise<User | null> {
+    return this.usersRepository.findOneByUsername(username);
+  }
```
구현을 보면 username에 해당하는 유저를 조회한다고 보이는데 findUserByUsername같은 메서드명은 어떤가요??

```suggestion
  async findUserForAuth(username: string): Promise<User | null> {
    return this.usersRepository.findOneByUsername(username);
  }
```

dlehdhun replied:
username으로 유저를 찾는 메서드라서 더 직관적인 것 같아요. 감사합니다!

rrgks6221 commented on src/users/users.service.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  // 비밀번호 확인용 유저 조회
+  async findUserForAuth(username: string): Promise<User | null> {
+    return this.usersRepository.findOneByUsername(username);
+  }
+
+  // 유저 생성
+  async createUser(userData: Prisma.UserCreateInput): Promise<userData> {
+    return this.usersRepository.createUser(userData);
+  }
```
두 메서드가 서로다른 user interface를 리턴해야하는 이유가 있을까요??
특별한 이유가 없다면 동일한 유저 객체를 리턴하는게 좋을거같아요

1. 두 인터페이스를 계속 유지보수해줘야함
2. 사용처에서 어떤 유저 객체를 리턴하냐에 따라 추가 동작이 필요해질 수 있음

dlehdhun replied:
`findUserForAuth`는 사용자를 검증하기 위해 호출되는 메서드이기 때문에, 비밀번호 비교를 위해 hashedPassword가 필요합니다. 그래서 hashedPassword가 포함되어있는  user interface를 사용하고 있습니다.
반면 `createUser`는 회원 가입 이후 발급 로직에서 사용되며, hashedPassword를 사용하지 않습니다. 그래서 hashedPassword를 제거 한 user interface를 사용하고 있습니다. 혹시 이런 상황에서는 어떤 방식으로 구현하는 것이 좋을까요?

rrgks6221 replied:
이렇게 처리하신 주요 목적이 hashedPassword를 숨겨야하기 떄문으로 보이는데요

서버 내부 로직(service 같은 레이어)에서는 hashedPassword가 서로 오가도 된다고 생각하고 controller에서 응답하기 전에만 응답 객체로 잘 감싸 은닉만 하면 된다고 생각하긴 합니다.

rrgks6221 commented on src/redis/redis.service.ts:
redis.service가 auth 디렉토리 하위에 위치해있는데 굉장히 일반화된 인터페이스를 제공하기 떄문에 prisma와 같이 상위 디렉토리로 이동시켜 다른 모듈에서도 사용했을 때 어색하지 않게 조정하는것도 좋을거같아요

dlehdhun replied:
말씀하신 대로 Redis가 다른 모듈에서도 활용될 수 있기 때문에 상위 디렉토리로 이동하는 게 더 적절한 것 같습니다!

dlehdhun committed:
refactor: (#5) salt rounds에 클래스 속성 사용 및 주석 제거

rrgks6221 commented on src/auth/token/token.service.ts:
```diff
@@ -0,0 +1,26 @@
+import { ConfigService } from '@nestjs/config';
+
+@Injectable()
+export class TokenService {
```
사람마다 관점이 다를 수 있는 부분인데요
TokenService가 정의된데에 비해서 역할이 적다고 느껴지는 부분이 있어요

제가 생각하는 tokenService의 역할과 책임: 토큰(로그인 세션 관련 토큰)에 대한 비즈니스 로직과 영속성(데이터베이스)에 대한 책임을 가짐

지금 구조는 token이 생성되거나 회수(로그아웃)할 때 redis와 싱크를 맞추는 구조로 파악했는데요.
아래 부분을 봤을 때 tokenService에서 redis에 관련된 처리까지 함께 하는것에 대해서는 어떻게 생각하시나요??
1. 토큰이 생성됨과 갱신 등등에 대한 처리에 후속조치로 redis 스토어에 데이터를 저장 및 삭제 수정하는 동작이 필수적으로 이루어져야하는 부분
2. token과 관련된 redis 키를 만드는 로직이 상위 서비스에서 처리되고 있는 부분

두 클래스를 합친다면 TokenService의 주요 동작은 아래처럼 될거같아요

1. 엑세스 토큰 생성
2. 리프래시 토큰 생성
3. 토큰 회수(로그아웃같은)
4. rotate(리프레시토큰을 통한 엑세스 토큰 재발급)
5. validate

dlehdhun replied:
저도 `TokenService`의 역할이 현재는 단순하다고 느꼈는데, 말씀해주신 것처럼 Redis 관련 처리를 `TokenService`내부에서 함께 관리한다면 전체 흐름을 한 곳에서 다룰 수 있어 유지보수 측면에서 더 효율적일 것 같습니다. 또한 Redis 처리 누락과 같은 실수도 줄일 수 있을 것 같아 좋은 방법 같습니다!

rrgks6221 commented:
충돌이 있는데 해결되면 좋을거같습니다.
이전 PR에서 봤던거같은 코드들이 있는데 중복되게 리뷰하지 않도록 베이스브랜치를 한번 땡겨오면 리뷰하는데 도움이 될거같아요

jungyoonho128 commented on src/auth/cookie/cookie.service.ts:
```diff
@@ -0,0 +1,28 @@
+    res.cookie(this.refreshName, refreshToken, {
+      httpOnly: true,
+      secure: this.configService.get<string>('COOKIE_SECURE') === 'true',
+      sameSite: 'lax',
```
CookieService에서 httpOnly, sameSite, path, secure 옵션이 반복되고 있는데 공통 cookieOptions로 묶으면 코드가 더 간결해질 것 같습니다. 또한 쿠키 이름과 maxAge를 .env에서 관리하면 환경별 유연성이 높아질 것 같습니다!

dlehdhun replied:
`cookieOptions`를 묶어서 사용하면 코드 가독성과 유지보수 측면에서 확실히 좋아질 것 같네요. 쿠키 이름과 maxAge도 환경 변수로 관리해서 반영해 보겠습니다!

jungyoonho128 commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,93 @@
+    });
+    if (existsNickname) {
+      throw new ConflictException('이미 사용 중인 닉네임입니다.');
+    }
```
현재 signup에서 username과 nickname 존재 여부를 각각 checkUserExists로 두 번 호출하고 있는데,
한 번의 DB 호출이나 트랜잭션으로 동시에 확인할 수 있으면 쿼리 수를 줄여 성능을 개선할 수 있을 것 같습니다.

dlehdhun replied:
한 번에 확인하거나 묶어서 호출하는 방법은 미처 생각하지 못했네요. 해당 방향으로 개선해 보겠습니다!

jungyoonho128 commented on src/auth/strategies/refresh-token.strategy.ts:
```diff
@@ -0,0 +1,36 @@
+    private readonly redisService: RedisService,
+  ) {
+    super({
+      jwtFromRequest: (req: Request) => {
+        const refreshToken = req.cookies?.['refreshToken'] as string | undefined;
+        return refreshToken ?? null;
```
CookieService와 RefreshTokenStrategy에서 refreshToken 쿠키 이름을 코드에 직접 작성하는 대신 환경변수(REFRESH_TOKEN_COOKIE)로 관리하면 코드 일관성과 환경별 유연성이 높아집니다. 또한 쿠키 존재 여부 체크를 추가하면 안전성이 향상될 것 같습니다!

dlehdhun replied:
쿠키 존재 여부 체크 부분은 생각하지 못했는데, 알려주셔서 감사합니다! 환경변수에 저장하는것도 코드 일관성과 환경별 유연성측면에서 좋은것 같아요

jungyoonho128 commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,93 @@
+    const { sub, nickname } = userData;
+    const accessToken = this.tokenService.createAccessToken({ sub, nickname });
+    return { accessToken };
+  }
```
AuthService의 refreshToken 메서드는 현재 동기 함수로 JWT 생성만 하고 있는데 향후 refreshToken 검증이나 Redis 조회 등 비동기 로직이 추가될 가능성이 있다면 미리 async 구조로 설계하면 확장성과 일관성을 유지할 수 있을 것 같습니다!

dlehdhun replied:
확장성과 일관성을 위해서 async 구조로 하면 좋을 것 같네요. 적용해보겠습니다!

jungyoonho128 commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,93 @@
+
+  // 로그인
+  async login(userData: userData): Promise<TokenDto> {
+    const paylode: JwtPayload = {
```
payloade 오타인 것 같습니다

dlehdhun committed:
refactor: (#5) merge 충돌 해결 및 Redis 모듈 이동, 토큰 저장 로직을 TokenService로 이동

- develop 브랜치 pull 후 발생한 merge 충돌 해결
- redis 폴더를 상위 디렉토리로 이동
- AuthService에서 처리하던 refresh 토큰 Redis 저장 로직을 TokenService로 이동

dlehdhun committed:
refactor: (#5) merge 충돌 해결 및 Redis 모듈 이동, 토큰 저장 로직을 TokenService로 이동

- develop 브랜치 pull 후 발생한 merge 충돌 해결
- redis 폴더를 상위 디렉토리로 이동
- AuthService에서 처리하던 refresh 토큰 Redis 저장 로직을 TokenService로 이동

Kimsoo0119 commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,83 @@
+      hashedPassword,
+    });
+
+    return this.login(newUser);
```
signup 기능이 현재 너무 많은 역할을 담당하고 있는 것 같습니다.
회원가입 성공 시 로그인까지 처리하고자 하신 의도는 이해되지만,
회원가입(signup)과 로그인(login)은 서로 다른 책임을 가진 기능이므로 분리하는 것이 좋다고 생각합니다.

예를 들어, signup API에서는 회원가입 결과(예: 201 Created)만 반환하고,
이후 클라이언트에서 해당 응답을 기반으로 login을 호출하는 흐름이 더 명확하고 유지보수에도 유리하지 않을까 생각합니다.

dlehdhun replied:
피드백 감사합니다! 자동 로그인은 서버에서 바로 처리하는 방식만 생각했는데, 회원가입 성공 후 클라이언트에서 로그인 API를 별도로 호출하는 방법도 가능하군요 프론트랑 상의해보겠습니다!

dlehdhun committed:
refactor: (#5) 회원가입 중복 검사 로직 단일 DB 조회로 변경 및 관련 구조 개선

- users.repository: findFirst로 username/nickname 동시 중복 검사
- token.module: RedisModule 추가
- accessToken 재발급 시 payload 누락 보완

dlehdhun committed:
refactor: (#5)  CookieService 제거 및 회원가입 자동 로그인 로직 제거

- 쿠키 설정/삭제 로직을 AuthController 내부 private 메서드로 통합
- 서비스 계층에서 Prisma DTO 대신 커스텀 DTO로 변경
- 회원가입 시 토큰 발급 로직 제거 → 단순 201 Created 응답 반환
